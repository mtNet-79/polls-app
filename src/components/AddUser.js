import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { connect } from "react-redux";
import { handleAddUser } from "../actions/addUser";
// WHEN PAGE OPENS RENDER NEW USER OBJ WITH NEW RANDOM IMAGE
// IF USER SUBMITS NAME AND PASSWORD ADD TO THE OBJECT THEN SUBMIT THE OBJECT TO DISPATCH
const LogIn = (props) => {
  const [userName, setUserName] = useState("");
  const [fullName, setFullName] = useState("");
  const [password, setPassword] = useState("");
  const [formReady, setFormReady] = useState(false);
  const navigate = useNavigate();
  const { dispatch, image, users } = props;
  

  const checkForm = (e) => {
    if (e.target.id === "fullName") setFullName(e.target.value);
    if (e.target.id === "userName") {
      Object.keys(users).forEach((userId) => {
        userName === userId 
            ? alert('This user name is already taken.')
            : setUserName(e.target.value);
      })
      setUserName(e.target.value);
    }
    else if (e.target.id === "password") setPassword(e.target.value);
    if (fullName !== "" && password !== "" && password.length > 5)
      setFormReady(true);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    // alert('what');
  
   
      userName 
        ? dispatch(handleAddUser({fullName, image, password, userName})) 
        : dispatch(handleAddUser({fullName, image, password, users}))    
    navigate("/");
  };

  //TODO: create random image

  return (
    <div className="container">
      <img src={image} alt="description" className="avatar" />
      <div>
        <form onSubmit={handleSubmit} className="add-new">
          <label htmlFor="fullName" className="form-label center">
            Full name:
          </label>
          <input
            placeholder="Full Name"
            type="text"
            name="fullName"
            id="fullName"
            value={fullName}
            onChange={checkForm}
          />
          <label htmlFor="fullName" className="form-label center">
            User name:
          </label>
          <input
            placeholder="User Name"
            type="text"
            name="userName"
            id="userName"
            value={userName}
            onChange={checkForm}
          />
          <label htmlFor="password" className="form-label center">
            Password
          </label>
          <input
            placeholder="Password"
            type="text"
            name="password"
            id="password"
            value={password}
            onChange={checkForm}
          />
          <button
            type="submit"
            className="btn add-new-btn"
            disabled={!formReady}
          >
            Add New User
          </button>
        </form>
      </div>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }, { image }) => ({
  authedUser,
  image,
  users,
});
export default connect(mapStateToProps)(LogIn);
