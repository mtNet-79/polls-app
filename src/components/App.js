import { handleInitialData } from "../actions/shared";
import PollPage from "./PollPage";
import { useState, useEffect, Fragment } from "react";
import Dashboard from "./Dashboard";
import { connect } from "react-redux";
import Nav from './Nav';
import LogIn from "./LogIn";
import Leaderboard from "./Leaderboard";
import AddUser from "./AddUser";
import LoadingBar from "react-redux-loading-bar";
import { Routes, Route } from "react-router-dom";
import CreatePoll from "./CreatePoll";
import { useLocation, useNavigate, useParams } from "react-router-dom";

const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};

const App = (props) => {
  console.log("app props: ", props);
  const [image, setImage] = useState("");
  useEffect(() => {
   
    const generateRandomAvatar = () => {
      fetch("https://avatar-endpoint.herokuapp.com/api/").then((res) => {
        res.blob().then((blobRes) => {
          const data = blobRes;
          const urlCreator = window.URL || window.webkitURL;
          const urlObj = urlCreator.createObjectURL(data);
          setImage(urlObj);
          console.log("trigged in generateavatar");
          props.dispatch(handleInitialData());
        });
      });
    }

    const alreadyLoggedIn = () => {
      console.log("triggerd at already")
      props.dispatch(handleInitialData());
    }
    
    props.authedUser === null 
      ? generateRandomAvatar() 
      : alreadyLoggedIn();
  }, []);



  return (
    <Fragment>
      <LoadingBar />       
        {props.loading ? null : (   
            <div className="container">              
              {props.location.pathname === "/add-user" 
                ? <AddUser image={image} />
                : props.authedUser
                  ? (
                    <Fragment>                   
                      <Nav />
                      <Routes>
                        <Route path="/" exact element={<Dashboard />} />                        
                        <Route path="/login" element={<LogIn image={image} />} />
                        <Route path="/new" element={<CreatePoll />} />
                        <Route path="/leaderboard" element={<Leaderboard />} />
                        <Route path="/questions/:pid" element={<PollPage image={image}/>} />
                      </Routes>
                    </Fragment>   
                  )           
                  : <LogIn image={image} />                  
              }            
            </div>
        )}
    </Fragment>
  );
};

const mapStateToProps = ({  authedUser, loading, users }, props) => {  
  // const keys = Object.keys(users);
  const { location } = props.router;
  return {
    loading: loading,
    authedUser,
    avatarURL: users[authedUser]?.avatarURL,
    location,
  }
    
    
};
  

export default withRouter(connect(mapStateToProps)(App));
