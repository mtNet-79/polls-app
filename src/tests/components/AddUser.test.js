import App from "../../components/App";
import { renderWithProviders } from "./test-utils";
import { screen, waitFor } from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import AddUser from "../../components/AddUser";

test('paints form and has submit feature', async () => {
    const user = userEvent.setup();
    renderWithProviders(<App />)
    await waitFor(
        () =>
          expect(
            screen.getByRole("link", { name: /add new user/i })
          ).toBeInTheDocument(),
        {
          timeout: 2500,
        }
      );

      //add user link is present
  const addNewUserBtn = screen.getByRole("link", { name: /add new user/i });
  expect(addNewUserBtn).toBeInTheDocument();

  await user.click(addNewUserBtn)

  //get all form elements

  const fullName = screen.getByRole('textbox', {name: /full name/i})
  const userName = screen.getByRole('textbox', {name: /user name/i})
  const password = screen.getByRole('textbox', {name: /password/i})

  const submitBtn = screen.getByRole('button', {name: /add new user/i})

  expect(submitBtn).toBeDisabled()

  
  await user.type(fullName, 'Michael Thornton');
  expect(fullName.value).toBe('Michael Thornton')
  await user.type(password, 'someT')

  expect(submitBtn).toBeDisabled()

  await user.type(password, 'ext')  
  //user name not required
  expect(userName.value).toBe('')
  expect(submitBtn).not.toBeDisabled()

  //optional user name input test
  await user.type(userName, 'dedalus')
  expect(userName.value).toBe('dedalus')

  await user.click(submitBtn)

  //takes user to dashboard after entering information

  //load mock api data is set to resolve at timeout: 1000
const header = await screen.findByText(/new polls/i)
expect(header).toBeInTheDocument()



})

test('snapshot', async () => {

    const user = userEvent.setup();
    const {container} = renderWithProviders(<App />)
    // const tree = renderer.create(<AddUser />).toJSON();
    await waitFor(
        () =>
          expect(
            screen.getByRole("link", { name: /add new user/i })
          ).toBeInTheDocument(),
        {
          timeout: 2500,
        }
      );

      //add user link is present
  const addNewUserBtn = screen.getByRole("link", { name: /add new user/i });
  expect(addNewUserBtn).toBeInTheDocument();

  await user.click(addNewUserBtn)

  //get all form elements

  expect(screen.getByRole('textbox', {name: /full name/i})).toBeInTheDocument()

  const submitBtn = screen.getByRole('button', {name: /add new user/i})

  expect(submitBtn).toBeDisabled()

  expect(container).toMatchSnapshot()

})