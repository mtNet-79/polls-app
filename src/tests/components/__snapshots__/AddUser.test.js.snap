// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    style="direction: ltr;"
  >
    <div
      class=""
      style="width: 100%; transition: width 200ms linear 0s; will-change: width, opacity; height: 3px; background-color: red; position: absolute; opacity: 1;"
    />
    <div
      style="display: table; clear: both;"
    />
  </div>
  <div
    class="container"
  >
    <div
      class="container"
    >
      <img
        alt="description"
        class="avatar"
      />
      <div>
        <form
          class="add-new"
        >
          <label
            class="form-label center"
            for="fullName"
          >
            Full name:
          </label>
          <input
            id="fullName"
            name="fullName"
            placeholder="Full Name"
            type="text"
            value=""
          />
          <label
            class="form-label center"
            for="userName"
          >
            User name:
          </label>
          <input
            id="userName"
            name="userName"
            placeholder="User Name"
            type="text"
            value=""
          />
          <label
            class="form-label center"
            for="password"
          >
            Password
          </label>
          <input
            id="password"
            name="password"
            placeholder="Password"
            type="text"
            value=""
          />
          <button
            class="btn add-new-btn"
            disabled=""
            type="submit"
          >
            Add New User
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
